src/openloco/windows/stationwnd.cpp

```diff
@@ -1,55 +1,660 @@
+            make_widget({ 3, 125 }, { 221, 10 }, widget_type::wt_13, 1),
+            widget_end(),
+        };
+
+        static window_event_list events;
 
-        const auto station = stationmgr::get(w.number);
-        int16_t y = 0;
-        for (int i = 0; i < 32; i++)
+        // 0x0048EC3B
+        static void prepare_draw(window* self)
         {
-            auto& cargo = station->cargo_stats[i];
-            if (!cargo.empty())
+            common::prepare_draw(self);
+
+            self->widgets[widx::scrollview].right = self->width - 4;
+            self->widgets[widx::scrollview].bottom = self->height - 14;
+
+            self->widgets[widx::status_bar].top = self->height - 12;
+            self->widgets[widx::status_bar].bottom = self->height - 3;
+            self->widgets[widx::status_bar].right = self->width - 14;
+
+            common::repositionTabs(self);
+        }
+
+        // 0x0048ED24
+        static void draw(window* self, gfx::drawpixelinfo_t* dpi)
+        {
+            self->draw(dpi);
+            common::drawTabs(self, dpi);
+        }
+
+        // 0x0048EE1A
+        static void on_mouse_up(window* self, widget_index widgetIndex)
+        {
+            switch (widgetIndex)
+            {
+                case common::widx::caption:
+                    common::renameStationPrompt(self, widgetIndex);
+                    break;
+
+                case common::widx::close_button:
+                    WindowManager::close(self);
+                    break;
+
+                case common::widx::tab_station:
+                case common::widx::tab_cargo:
+                case common::widx::tab_cargo_ratings:
+                    common::switchTab(self, widgetIndex);
+                    break;
+            }
+        }
+
+        // 0x0048EE97
+        static void on_resize(window* self)
+        {
+            // Call to sub_48E32C has been deliberately omitted.
+
+            self->set_size(gfx::ui_size_t(340, 208), gfx::ui_size_t(340, 208));
+        }
+
+        // 0x0048EE4A
+        static void get_scroll_size(window* self, uint32_t scrollIndex, uint16_t* scrollWidth, uint16_t* scrollHeight)
+        {
+            auto station = stationmgr::get(self->number);
+            *scrollHeight = 0;
+            for (uint8_t i = 0; i < 32; i++)
+            {
+                if (station->cargo_stats[i].origin != 65535)
+                    *scrollHeight += 10;
+            }
+        }
+
+        // 0x0048EE73
+        static void tooltip(FormatArguments& args, ui::window* window, widget_index widgetIndex)
+        {
+            args.push(string_ids::tooltip_scroll_ratings_list);
+        }
+
+        // 0x0048EF02
+        static void draw_rating_bar(window* self, gfx::drawpixelinfo_t* dpi, int16_t x, int16_t y, uint8_t amount, colour_t colour)
+        {
+            gfx::fill_rect_inset(dpi, x, y, x + 99, y + 9, self->colours[1], 48);
+
+            uint16_t rating = (amount * 96) / 256;
```

Comment:
This is missing a whole bunch of additional code that makes the rating bar flash

Reply:
I didn't think that code is ever called. As you are Bit Testing and Resetting the 32nd bit of the colour which is always zero as the inputs are 13, 16 or 27. I've looked in the original game and don't see the rating bars flash at any point. Unless I've overlooked something?

Reply:
I suppose it could be code that's reused from RCT. The bars for guests stats do blink sometimes there.